src/compiler/utilities.ts

```diff
@@ -4298,6 +4298,13 @@ namespace ts {
         return declaration.name || nameForNamelessJSDocTypedef(declaration);
     }
 
+    /** @internal */
+    export function isNamedDeclaration(node: Node): node is NamedDeclaration & { name: DeclarationName } {
+        const name = (node as NamedDeclaration).name;
+        Debug.assert(!name || isDeclaration(node) && node.name === name, "A 'name' property should always be a DeclarationName.");

```

Comment:
this really a hot path, i would rather not do additional work. this is rather simple check. does it have a `name` property or not.